src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g

```diff
@@ -1424,7 +1424,7 @@ unaryExpressionNotPlusMinus
     ;
 
 typeCastParameters
-    : classTypeSpec[true] (BAND^ classTypeSpec[true])*
+    : classTypeSpec[true] (b:BAND {#b.setType(TYPE_EXTENSION_AND);} classTypeSpec[true])*
```

Comment:
`^` is removed because there is no need for a hierarchy. `&` is just a separator for a list of types.
Consider the following code:
```
class Foo {
    Object o = (Foo & Cloneable & Cloneable & Cloneable) null;
}
```
For now the AST is
```
--TYPECAST -> ( [2:15]
   |--BAND -> & [2:44]
   |   |--BAND -> & [2:32]
   |   |   |--BAND -> & [2:20]
   |   |   |   |--TYPE -> TYPE [2:16]
   |   |   |   |   `--IDENT -> Foo [2:16]
   |   |   |   `--TYPE -> TYPE [2:22]
   |   |   |       `--IDENT -> Cloneable [2:22]
   |   |   `--TYPE -> TYPE [2:34]
   |   |       `--IDENT -> Cloneable [2:34]
   |   `--TYPE -> TYPE [2:46]
   |       `--IDENT -> Cloneable [2:46]
   |--RPAREN -> ) [2:55]
   `--LITERAL_NULL -> null [2:57]
```
Note for the 'ladder' of `BANDs`. It is hard to iterate and it just make no sense. 
By the way, for the `extends` there is no hierarchy either. For the code:
```
class Foo<T extends Foo & Cloneable & Cloneable & Cloneable> {
}
```
the AST is
```
--TYPE_UPPER_BOUNDS -> extends [1:12]
   |--IDENT -> Foo [1:20]
   |--TYPE_EXTENSION_AND -> & [1:24]
   |--IDENT -> Cloneable [1:26]
   |--TYPE_EXTENSION_AND -> & [1:36]
   |--IDENT -> Cloneable [1:38]
   |--TYPE_EXTENSION_AND -> & [1:48]
   `--IDENT -> Cloneable [1:50]
```

Reply:
> & is just a separator for a list of types.

Comma separated lists like arrays and list of implements in class also don't have a hierarchy in AST. so I am fine with leaving it like this.

Reply:
@pbludov 
> By the way, for the extends there is no hierarchy either. 

Please add an example to grammar tests, I am not seeing one.

Reply:
done